ffmpeg-plugin/mcm_audio_tx.c

```diff
@@ -114,29 +117,50 @@ static int mcm_audio_write_header(AVFormatContext* avctx)
 static int mcm_audio_write_packet(AVFormatContext* avctx, AVPacket* pkt)
 {
     McmAudioMuxerContext *s = avctx->priv_data;
+    size_t frame_size = s->tx_handle->frame_size;
     const uint8_t *data = pkt->data;
-    mcm_buffer *buf = NULL;
     int size = pkt->size;
     int len, err = 0;
 
     while (size > 0) {
-        buf = mcm_dequeue_buffer(s->tx_handle, -1, &err);
-        if (buf == NULL) {
-            av_log(avctx, AV_LOG_ERROR, "Dequeue buffer error %d\n", err);
-            return AVERROR(EIO);
+        char *dest;
+
+        if (!s->unsent_mcm_buf) {
+            s->unsent_mcm_buf = mcm_dequeue_buffer(s->tx_handle, -1, &err);
```

Comment:
I agree it looks suspicious. But there is no major problem. Let me explain that. We use this memif connection to communicate with Media Proxy only, which controls the connection. If the connection is closed, `mcm_dequeue_buffer()` returns. The flow stays blocked here until the memif library provides an empty buffer to fill. And this is totally fine because if the call is timed and a timeout occurs, the flow will have to break transmission and that's not what we want to happen.

Let's keep this approach as it is for now since it is working.

Reply:
It is more like a question if you have multiple audio buffers won't this block any other possible operations ongoing (so how the multithreading is handled). Also getting buffer if the queue is full will not work as intended, and should we do an audio frame skip then or wait?

Reply:
For sure we can't skip anything. That's why infinite waiting is meaningful here. If a skip happens, the queue capacity has to be increased.

When it comes to multiple audio streams, every stream is processed by FFmpeg in a separate thread. And every connection to MCM creates a separate memif queue which is also processed in separate threads. So, the implementation is thread-safe. We don't use any shared data here between the audio streams.